### PR TITLE
meshery: Fix nil messages and generate verbose logs

### PR DESCRIPTION
**Pull request by kanishkarj**
Signed-off-by: kanishkarj <kanishkarj@hotmail.com>

#### Description

This PR fixes the nil dereferencing error messages that are logged right after login. These errors occur when the K8s config is `nil`. `nil` checking added.
This PR also adds a bit more verbose logging while logging in as to figure out in case of login failure.

#### Checklist

- [x] [DCO](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin) signed.
- [x] Title of the PR starts with area name, e.g. `meshery:`, `mesheryctl:`, `ui:`.

### CODE REVIEW

**Comment by kanishkarj** [comment]
@kushthedude , try running this PR on your system, maybe more verbose logs could help.

**Comment by kushthedude** [comment]
@kanishkarj Same error logs as before see here 😓  
```sh
 ~/meshery   892 ⍟1  make run-local          SIGINT(2) ↵  11559  19:03:17
cd cmd; go clean; rm meshery; go mod tidy; go build -tags draft -a -o meshery; \
	SAAS_BASE_URL=https://meshery.layer5.io \
	PORT=9081 \
	DEBUG=true \
	ADAPTER_URLS="mesherylocal.layer5.io:10000 mesherylocal.layer5.io:10001 mesherylocal.layer5.io:10002 mesherylocal.layer5.io:10003 mesherylocal.layer5.io:10004 mesherylocal.layer5.io:10008" \
	./meshery; \
	cd ..
rm: meshery: No such file or directory
INFO[0000] Using '/Users/kush/.meshery/config' to store user data
INFO[0000] Using '/Users/kush/.kube' as the folder to look for kubeconfig file
INFO[0000] Log level: debug
INFO[0000] Starting Server listening on :9081
ERRO[0010] error in getting the token, http: named cookie not present
INFO[0010] Session not found
DEBU[0022] token : eyJhY2Nlc3NfdG9rZW4iOiJleUpoYkdjaU9pSlNVekkxTmlJc0ltdHBaQ0k2SW5CMVlteHBZenBsT0dWbU5ERmpNeTFpWldWbUxUUmlZakV0T0dVNE1DMHpOakExTVRZeU4yTTJNakVpTENKMGVYQWlPaUpLVjFRaWZRLmV5SmhkV1FpT2x0ZExDSmpiR2xsYm5SZmFXUWlPaUp0WlhOb1pYSjVMV05zYjNWa0lpd2laWGh3SWpveE5UZzNOek01TURjeUxDSmxlSFFpT250OUxDSnBZWFFpT2pFMU9EYzNNelUwTnpJc0ltbHpjeUk2SW1oMGRIQnpPaTh2YldWemFHVnllUzVzWVhsbGNqVXVhVzh2YUhsa2NtRXZJaXdpYW5ScElqb2lOVFprTldabFptWXROVGRtTkMwMFlUbGlMV0ZpTTJJdFpqVm1ZemMwWW1aak4yVTBJaXdpYm1KbUlqb3hOVGczTnpNMU5EY3lMQ0p6WTNBaU9sc2liM0JsYm1sa0lpd2liMlptYkdsdVpTSmRMQ0p6ZFdJaU9pSmhNMVo2WVVoU2IxcFhVakZhUjFVOU9sb3liREJoU0ZacEluMC5HbW55TmkweXNhTVhrSERDTmdXd05NNGlIUFBvc1FCR2tLaVgyd2E2TTlhSFpqdDNTMUM2TzBEdEY3U1ByNXJaQXZxd1dlTUUwUWhpajlrUTM0ODlrcEItWUMyUFlxMldoUW85Ym1iaTlxMlhlS0x3US1Jb2JXWVpFdDFVaXhiSG1WdkNqLUFvZmpwazJlOHhpSzIxSnlMdkFvUEMyb2sycG15U2YyOWZZQ05QbmZmVGlua042dndxTW1wWGdOSzc3amdWdDZTakdKYnl2cGxFY2ZSTHpRd3huXzNsckFmTEsxOUFFeEdlRjlVVUxpemlVdkp0Zkx0ZUFCbk0ydGJLXzJuOXJxU21BYnR1NnNxeENQNHUxT1Ntek9fUlJrajF0aUtuNE55cEdHQnRYQUtPYzFSV2Q2WTNvSUUzRkttcUgtZDg2RzNUR2FFZldoa25RWkdjQnFtNVJ5MlhpVnk1SGd0QTlSdkZFUUlqaUllSk13LV8xUXlwdl9ROVBTeEhLMjRIUUxqTFlRcmpqdG9VWXNFeE96ZU5uOUVsQlIzQUhzZHRQM1VMT3pEOG5oM21oSGhlaU1CWVl2VXlVQ1pYZ3F1Wi1GMXpjdEx5Q29QejB2OXh3bzBVT01CaER6TEp6MkVOM1IzcWo0ZHRycnoyNlJleHBWNGY4N3JlcFpNZW4yVms0c1ZiNnZNTkQxcXR5RTZRMHNFNUFRQkRIa3BRVElyTUpnNXFsaU5uVmJab2FRWHBqREtHUS1nZnZSQUJ3MHpxN0tneEo5V2xLMEdhT19xa3RkbmQxdk5JU2d6eUpGVzQ2Q0VtZm9hZkoyR3hETkZJLU1qRTduQnEyVU10T01SaFBvdWhjSEctNzJuc3ZFOW1OcVVUeWcyYnFvdkJrM2xlYUEzdWtIbyIsInRva2VuX3R5cGUiOiJiZWFyZXIiLCJyZWZyZXNoX3Rva2VuIjoiaDNvYVZ1d1l1aVE4SjFSYkwyMHZaYnM4SlhGcThjZ3BwTUVaaVlfdk9VTS5qOVdVN2Ntd0dSUEpmRm51WklQT2xMdnpHcHBVZXBtM1ZRVHpLM3JMRFhzIiwiZXhwaXJ5IjoiMjAyMC0wNC0yNFQxNDozNzo1MS4yNzg5ODk0NzVaIn0
ERRO[0024] error validating token : Token is not valid yet
INFO[0024] Token validation error : Token is not valid yet
ERRO[0025] error validating token : Token is not valid yet
ERRO[0025] Validation of refreshed token failed : Token is not valid yet
```

**Comment by kanishkarj** [comment]
The issue is similar to (this)[https://github.com/nokia/kong-oidc/issues/59]. I feel the server time and your  system time are not in sync or something. Doesn't make sense a bit, but look a the error "token is not valid yet"? like what.

**Comment by kanishkarj** [comment]
@leecalcote I noticed those error logs which occur due to a panic (due to nil dereferencing) [here](https://github.com/layer5io/meshery/pull/886#issuecomment-618071356).  I have fixed that here.

**Comment by kushthedude** [comment]
> The issue is similar to [this](https://github.com/nokia/kong-oidc/issues/59). I feel the server time and your system time are not in sync or something. Doesn't make sense a bit, but look a the error "token is not valid yet"? like what.

I have not used hydra, dont know much about it!

This seems strange because

**Comment by kanishkarj** [comment]
@kushthedude Try the running it now now.

**Comment by kushthedude** [comment]
There is no issue with my system & server time.

@kanishkarj Alas, still not working. Here are the logs : 
```sh
cd cmd; go clean; rm meshery; go mod tidy; go build -tags draft -a -o meshery; \
	SAAS_BASE_URL=https://meshery.layer5.io \
	PORT=9081 \
	DEBUG=true \
	ADAPTER_URLS="mesherylocal.layer5.io:10000 mesherylocal.layer5.io:10001 mesherylocal.layer5.io:10002 mesherylocal.layer5.io:10003 mesherylocal.layer5.io:10004 mesherylocal.layer5.io:10008" \
	./meshery; \
	cd ..
INFO[0000] Using '/Users/kush/.meshery/config' to store user data
INFO[0000] Using '/Users/kush/.kube' as the folder to look for kubeconfig file
INFO[0000] Log level: debug
INFO[0000] Starting Server listening on :9081
ERRO[0015] error in getting the token, http: named cookie not present
INFO[0015] Session not found
DEBU[0027] token : eyJhY2Nlc3NfdG9rZW4iOiJleUpoYkdjaU9pSlNVekkxTmlJc0ltdHBaQ0k2SW5CMVlteHBZenBsT0dWbU5ERmpNeTFpWldWbUxUUmlZakV0T0dVNE1DMHpOakExTVRZeU4yTTJNakVpTENKMGVYQWlPaUpLVjFRaWZRLmV5SmhkV1FpT2x0ZExDSmpiR2xsYm5SZmFXUWlPaUp0WlhOb1pYSjVMV05zYjNWa0lpd2laWGh3SWpveE5UZzNOelUwTURRMkxDSmxlSFFpT250OUxDSnBZWFFpT2pFMU9EYzNOVEEwTkRZc0ltbHpjeUk2SW1oMGRIQnpPaTh2YldWemFHVnllUzVzWVhsbGNqVXVhVzh2YUhsa2NtRXZJaXdpYW5ScElqb2laR1prTmpNd016SXRabU13TUMwME5ESTJMV0ppTmpNdE5UQXlNR1V5WkdRME1qVm1JaXdpYm1KbUlqb3hOVGczTnpVd05EUTJMQ0p6WTNBaU9sc2liM0JsYm1sa0lpd2liMlptYkdsdVpTSmRMQ0p6ZFdJaU9pSmhNMVo2WVVoU2IxcFhVakZhUjFVOU9sb3liREJoU0ZacEluMC5ZOEpUUDlGT19aSmRDemJpMUxjUFdQNmZTSE5pQ05FTmptWGRZZE52R05tSkZnOEFxLTRPdVpiMDlEN0pOcVEwZzc1U1RhWERYVWxmakJfb19oU3RRT3pncnprNmkzTmkxYlEtYU1seHVyTGowQldYazliejFBd0J4MlpLSEtLV0VIMHZuQU9uNVFsdVZPa3pubXl6ck1zNm5lS2ZxaHZob0pCc28xTlltb1R3VFk5TGhiRl81aG1kUGMzellsY3VhNzdjc01aLWt6akVsNkxsVEM1a0JRVWtaT0RxYkZwTWdVNlJIellIUW1rSzVrdGJYMzJqcF9lSVZocGZOVUZqLWRmS1hpVk1oXzcxSF96WTB1a2YzZURnV0hFcGczOVAwT21wX3I2V1AxSC1pbXlWNzUxVHNXcWNSQ1BoREpXMmxaQmNEYnlPazVZeGE0MS1udnhYYlZGWHhWdjZId2dmVVZVMVhCZGtQd19icnlXVzRDQnRmcGNkcnphVzVlTG5uREdvcWtMR2xSQnB0U0tpa05MY1hyYnBGd2pDRFh6YnFCZ1JfcXVsZ2pvUkNDUVd4c1BsdjJhV0N3aEM2SFZ2T3hxSlJWOWQ1d09SWmZYRWxrZE40dzNDQlRWNmJhR3ZmSFBTdHZxblg5aUFjczJ4aW9CLW5VZHV2bFp2XzNLRmxoejBuRjEtS3cxS1Z4ZFgtZkMzZV9yM2NoNnltcDVWY2ZrbnFnSkxBQjdTMFB6RjFOOHNNdkdKdjVvLURGWmt4X1lfWFNxdEt3eGJqTWFON0RTUE5zTkFYSXZhWUphbFhSdGoyd0ZPcnZsRHFhMmM5NlhqckYwMDcxa2dOdkdjX2FUd05zNUdseXdmVUg0TlBsOUVDYURTMFdBUklKQ0t4cHB4Mktpam91RSIsInRva2VuX3R5cGUiOiJiZWFyZXIiLCJyZWZyZXNoX3Rva2VuIjoiSlJsSldzT1M1OE9WR2VDMHJVQ2tWUlBVN2xKbU1TaUIzSnk1Nm5qR0hBTS4yRmI4ejNhaG5DdHRGTnBhWGwybGJBQ1NUamdIdDBuMi1aaElUdGlIeFhBIiwiZXhwaXJ5IjoiMjAyMC0wNC0yNFQxODo0NzoyNS4zOTAyNjYzMTFaIn0
DEBU[0027] Token issued at : 2020-04-24 23:17:26 +0530 IST 1587750446
DEBU[0027] Current system time : 2020-04-24 23:17:19.187289 +0530 IST m=+27.919497642
ERRO[0030] error validating token : Token is not valid yet
INFO[0030] Token validation error : Token is not valid yet
DEBU[0031] Token issued at : 2020-04-24 23:17:26 +0530 IST 1587750446
DEBU[0031] Current system time : 2020-04-24 23:17:22.455497 +0530 IST m=+31.187606799
ERRO[0031] error validating token : Token is not valid yet
ERRO[0031] Validation of refreshed token failed : Token is not valid yet
```

**Comment by kanishkarj** [comment]
@kushthedude @nveenjain this is an interesting behaviour, look at the time at which the token is issued by the server (the first time in logs) and the time at which the system is trying to authenticate (the second). The system is accessing the time 4 seconds before the time when the token was signed.

This is definitely not a time zone issue by the way because all tokens are signed and verified by using Unix time.

I simple hack that'll fix this would be to add an time offset of a minute or so. But if anyone has a solution to this issue let me know.

**Comment by kushthedude** [comment]
Try inquiring more about in hydra if some issue is already there addressing
this.

**Comment by kanishkarj** [comment]
First, I'll investigate if it's an issue with hydra. Because it seems to work on everyone else's system.

**Comment by kanishkarj** [comment]
@kushthedude check this, I guess we're facing a similar issue here : https://stackoverflow.com/questions/38495603/unix-timestamp-is-different-in-two-computers

**Comment by kushthedude** [comment]
That is an issue on hydra side, open an issue over there and see their
response.
Managing the distributed clock skew should never be the concern of
developer rather the service provider should look at it like tyk has
solved the issue.

**Comment by kushthedude** [comment]
See here https://github.com/TykTechnologies/tyk/issues/1670

**Comment by leecalcote** [comment]
@kushthedude is there a VM involved in your dev setup?

**Comment by kanishkarj** [comment]
I'm not sure if hydra will give any useful response. We are doing validation of the token using go-jwt. And considering the fact that the same authentication setup is working well on 5-6 of our systems, I don't think hydra is doing something wrong here. They are just signing the tokens.

@kushthedude, in the issue you point out at TYK, the solution they used is to provide a configuration variable to disable validation of 'issued at time', which is again a client side thing.

**Comment by kushthedude** [comment]
No there is no VM involved.
If hydra has such variable then tell me, we can disable the issue of time.
 Also, JWT gets expired after certain period of time, hence verification of
time is not a big deal in privacy or security of application.

**Comment by kushthedude** [comment]
Also when using any services, it should be the sole responsibility of
service provider that none of the clients gets any issue because of the
service. If they are getting, they should solve the problem or atleast give
them a reply.
Open an issue in hydra and point them to the issue, if the service is well
maintained they will give you an appropriate solution.
If not, then we should look for other solutions then using hydra service.

**Comment by leecalcote** [comment]
> No there is no VM involved. If hydra has such variable then tell me, we can disable the issue of time. Also, JWT gets expired after certain period of time, hence verification of time is not a big deal in privacy or security of application.
> […](#)

No variable/ configuration that I know of, but I haven’t looked. I’m asking about a VM, because clock drift is / was often an issue in virtual environments. 

Let’s do any further troubleshooting on a call.

**Comment by kushthedude** [comment]
I suspect why time issue is happening on my local server as I have set my
time according to my isp. If I had set up manually, I could understand the
issue.
Let's discuss more on call.

**Comment by kanishkarj** [comment]
@kushthedude, try out the latest commit please

File: handlers/mesh_ops_handlers.go [hunk 1]
```diff
@@ -108,9 +109,13 @@ func (h *Handler) addAdapter(ctx context.Context, meshAdapters []*models.Adapter
 		logrus.Debugf("Adapter already configured...")
 		return meshAdapters, nil
 	}
-
+	if prefObj.K8SConfig == nil {
+		err := fmt.Errorf("K8s config is nil")
```
Comment (kushthedude) [inline handlers/mesh_ops_handlers.go:113]: logrus.error can be used here directly.

File: models/meshery_remote_provider.go [hunk 2]
```diff
@@ -198,21 +198,21 @@ func (l *MesheryRemoteProvider) GetUserDetails(req *http.Request) (*User, error)
 func (l *MesheryRemoteProvider) GetSession(req *http.Request) error {
 	ts, err := l.GetToken(req)
 	if err != nil {
-		err = errors.Wrap(err, "Error: unable to get session")
-		logrus.Error(err)
+		err = fmt.Errorf("Session not found")
```
Comment (kushthedude) [inline models/meshery_remote_provider.go:201]: Same over here, `logrus.Error` can be used directly.

File: models/remote_auth.go [hunk 3]
```diff
@@ -196,7 +203,7 @@ func (l *MesheryRemoteProvider) GenerateKey(jwk JWK) (*rsa.PublicKey, error) {
 
 	der, err := x509.MarshalPKIXPublicKey(pk)
 	if err != nil {
-		logrus.Fatal(err)
+		logrus.Errorf("error mashalling PKIX, :", err.Error())
```
Comment (kushthedude) [inline models/remote_auth.go:206]: Use `logrus.Fatal`

File: models/remote_auth.go [hunk 4]
```diff
@@ -207,6 +214,7 @@ func (l *MesheryRemoteProvider) GenerateKey(jwk JWK) (*rsa.PublicKey, error) {
 
 	var out bytes.Buffer
 	if err := pem.Encode(&out, block); err != nil {
+		logrus.Errorf("error encoding jwk to pem, :", err.Error())
```
Comment (kushthedude) [inline models/remote_auth.go:217]: Use `logrus.Fatal`

File: models/remote_auth.go [hunk 5]
```diff
@@ -216,31 +224,57 @@ func (l *MesheryRemoteProvider) GenerateKey(jwk JWK) (*rsa.PublicKey, error) {
 func (l *MesheryRemoteProvider) VerifyToken(tokenString string) (*jwt.MapClaims, error) {
 	dtoken, err := l.DecodeTokenData(tokenString)
 	if err != nil {
-		logrus.Error("E1")
+		logrus.Error("error decoding token : %v", err.Error())
```
Comment (kushthedude) [inline models/remote_auth.go:227]: Use `logrus.Fatal`

File: models/remote_auth.go [hunk 6]
```diff
@@ -216,31 +224,57 @@ func (l *MesheryRemoteProvider) GenerateKey(jwk JWK) (*rsa.PublicKey, error) {
...
 		return nil, err
 	}
 	tokenString = dtoken.AccessToken
-	tokenUP, _, err := new(jwt.Parser).ParseUnverified(tokenString, jwt.MapClaims{})
+	tokenUP, x, err := new(jwt.Parser).ParseUnverified(tokenString, jwt.MapClaims{})
 	if err != nil {
-		logrus.Error("2")
+		logrus.Error("error parsing token (unverified) : %v", err.Error())
```
Comment (kushthedude) [inline models/remote_auth.go:233]: Use `logrus.Fatal`

File: models/remote_auth.go [hunk 7]
```diff
@@ -216,31 +224,57 @@ func (l *MesheryRemoteProvider) GenerateKey(jwk JWK) (*rsa.PublicKey, error) {
...
+	// TODO: Once hydra fixes https://github.com/ory/hydra/issues/1542
+	// we should rather configure hydra auth server to remove nbf field in the token
+	exp := int64(jtk["exp"].(float64))
+	if jwt.TimeFunc().Unix() > exp {
+		err := fmt.Errorf("Token has expired")
+		logrus.Errorf("error validating token : %v", err.Error())
+		return nil, err
+	}
+
+	// Enable only for Debugging
```
Comment (kushthedude) [inline models/remote_auth.go:251]: This block can be removed, not an important block. User can write manually for debugging or use the debugger.

File: models/remote_auth.go [hunk 8]
```diff
@@ -216,31 +224,57 @@ func (l *MesheryRemoteProvider) GenerateKey(jwk JWK) (*rsa.PublicKey, error) {
...
+	// iat := int64(jtk["iat"].(float64))
+	// tm := time.Unix(iat, 0)
+	// logrus.Debugf("Token issued at : %v %v", tm, iat)
+	// logrus.Debugf("Current system time : %v", time.Now())
+
 	keyJSON, err := l.GetJWK(kid)
 	if err != nil {
-		logrus.Error("3")
+		logrus.Error("error fetching JWK corresponding to token : %v", err.Error())
```
Comment (kushthedude) [inline models/remote_auth.go:259]: Use `logrus.Fatal`

File: models/remote_auth.go [hunk 9]
```diff
@@ -216,31 +224,57 @@ func (l *MesheryRemoteProvider) GenerateKey(jwk JWK) (*rsa.PublicKey, error) {
...
 		return nil, err
 	}
 	key, err := l.GenerateKey(keyJSON)
 	if err != nil {
+		logrus.Error("error generating Key from JWK : %v", err.Error())
```
Comment (kushthedude) [inline models/remote_auth.go:264]: Use `logrus.Fatal`

File: models/remote_auth.go [hunk 10]
```diff
@@ -216,31 +224,57 @@ func (l *MesheryRemoteProvider) GenerateKey(jwk JWK) (*rsa.PublicKey, error) {
...
+	// Verifies the signature
+	tokenParser := jwt.Parser{
+		SkipClaimsValidation: true,
+	}
+	token, err := tokenParser.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
 		return key, nil
 	})
 
 	if err != nil {
+		logrus.Errorf("error validating token : %v", err.Error())
```
Comment (kushthedude) [inline models/remote_auth.go:277]: Use `logrus.Fatal`

File: models/meshery_remote_provider.go (hunk 2 above)
Comment (kanishkarj) [reply models/meshery_remote_provider.go:201]: `fmt.Error` does not exist, only logrus has `Errorf`

Comment (kushthedude) [reply models/meshery_remote_provider.go:201]: Logurs.fatal as used to maintain consistency.

**Comment by kanishkarj** [comment]
@kushthedude Check now, at places I have used Info instead of Fatal as it seemed more appropriate.
@leecalcote I think the issue is fixed.

**Pull request merged**